### ISSUE TITLE
Support for ue4 .24

### ISSUE THREAD

**Issue by sonicphi** [PlayFab/gsdk#106]
I tried getting it to work on .24 but it doesnt compile, would be a nice addition support for older versions.

**Comment by dgkanatsios** [comment]
Thanks for reporting that! What errors are you seeing?

**Comment by sonicphi** [comment]
Well first of all it doesnt compile when you try to run VS with the plugin, also it complains about the module not existing when u try to open the project the regular way, even when you follow the tutorial steps.

**Comment by sonicphi** [comment]
If you want shoot me an email, its the same as my nickname @hotmail.com and we can have a discord chat or something similar.

**Comment by Hayden-W** [comment]
I have done some work to get the native UE4 plugin working on 4.25, Got it compiling but ran into an issue where the heartbeat from the MPS agent wasn't working correctly. (responses weren't being received / processed from MPS agent IIRC) 
I believe it may be to do with the fact that the http module SharedPtr's weren't thread safe and were being accessed from the heartbeat thread... maybe

[I believe that this commit fixes the issue I was seeing by making http shared ptrs thread safe by default](https://github.com/EpicGames/UnrealEngine/commit/30745eb3c6bcc16ec2f4bdd63fcc2baf66eebe06)

It's on my list to find a solution to get the native plugin working for 25,  might be able to lend a hand with this too if needed 👍

**Comment by sonicphi** [comment]
It wasnt compiling on .24 because of a pointer that was thread safe an the types were a miss match, somethign along the lines of "cant convert (xxxxx,1) to (xxxxx,0)", the number being 0 = unsafe and 1 = safe i think, it was an ENUM which i cant remember the name.

**Comment by Hayden-W** [comment]
> It wasnt compiling on .24 because of a pointer that was thread safe an the types were a miss match, somethign along the lines of "cant convert (xxxxx,1) to (xxxxx,0)", the number being 0 = unsafe and 1 = safe i think, it was an ENUM which i cant remember the name.

Yes that's what I encountered in 4.25, Changing the plugin to use the ptr type used by the HTTP module results in the weird heartbeat behaviour I saw above.

I believe that the git commit above contains the key to getting it to work w/ the plugin in it's current state, assuming you are compiling  the engine yourself.

to fix the issue without modifying the engine may require a lot of changes on the plugin side / or a separate plugin version for <4.26 engine versions.

**Comment by dgkanatsios** [comment]
We try to support the latest 2-3 versions, unfortunately. Any reasons prohibiting the use of a newer version than 4.24?

**Comment by sonicphi** [comment]
well, i was about to switch to .26 just so i could use the plugin, theres no reason per se besides going through the hassle of upgrading.

**Comment by JSaueressig88** [comment]
@Hayden-W I am currently running into this exact issue and haven't found a solution yet. Where did you end your pursuit here for this UE4.25 issue?

**Comment by JSaueressig88** [comment]
@dgkanatsios is UE4.25 not considered supported for this plugin either?

**Comment by dgkanatsios** [comment]
so far we've tested on 4.26/4.27. Unreal 4.25 was released about two years ago, any reason you can't use it?

**Comment by Hayden-W** [comment]
@JSaueressig88 
Hey mate, I didn't end up finishing the implementation in 4.25, Already had a working implementation of the CPP GSDK in a custom plugin that was doing the job.

We ended up updating to 4.27 at the start of the year and are now on the default UE plugin, I can probably dig up the code we used for 4.25 if required?

Cheers

**Comment by JSaueressig88** [comment]
> @JSaueressig88 Hey mate, I didn't end up finishing the implementation in 4.25, Already had a working implementation of the CPP GSDK in a custom plugin that was doing the job.
> 
> We ended up updating to 4.27 at the start of the year and are now on the default UE plugin, I can probably dig up the code we used for 4.25 if required?
> 
> Cheers

That sounds pretty great for trying out something new, I'd really appreciate it! :)

**Comment by Hayden-W** [comment]
@JSaueressig88 Would it be OK if I just emailed a zip of the code to you? If not then all good, Just don't have time to setup a repo right this sec, can get it to you over the weekend if that's the case 👍

**Comment by JSaueressig88** [comment]
> @JSaueressig88 Would it be OK if I just emailed a zip of the code to you? If not then all good, Just don't have time to setup a repo right this sec, can get it to you over the weekend if that's the case 👍

Yeah, for sure! Email: Jacob.David.Saueressig@gmail.com

**Comment by Hayden-W** [comment]
@JSaueressig88 

Hey mate, email was not cooperating last night, have created a repo here: https://github.com/Hayden-W/gsdk-ue-425

It's basically just the CPP version of the GSDK bundled with a C style communication layer to avoid conflicts with UE's weird memory stuff (Not 100% sure it was actually necessary but it works so ya know)

Everything you need should be in there, just include the GSDK module and it'll give you access to the basics of the GSDK implementation (Ready for players, session cookies, callbacks etc)

the gsdk version is kinda old, from like mid last year, but was working as of ~2 months ago

If you have any questions feel free to give me a shout :)

Cheers

**Comment by Malan-Tai** [comment]
@haydenstech Hey, I've found myself in the same situation as you described earlier : using UE4.25, there as a compile error in the GSDK module. I fixed it by removing the expected `ESPMode::ThreadSafe` from the GSDK source code. But when I laucnhed the server, I had a repeating "Failed to parse heartbeat" error message.

I found this thread and tried to modify the UE source code as you did in [this commit](https://github.com/EpicGames/UnrealEngine/commit/30745eb3c6bcc16ec2f4bdd63fcc2baf66eebe06) but the parse error remained.

I thus tried to use the repo you linked right above, but it looks very different to the current GSDK, and I therefore don't really know how to use it. Is there any chance you remember and could explain briefly either if there was anything more that needed change when modifying the UE source code, or how to use your custom plugin ?

Thanks in advance
